src/actions/ExchangeRateActions.js

```diff
@@ -17,79 +32,88 @@ export const updateExchangeRates = () => async (dispatch: Dispatch, getState: Ge
-  // Promise.allSettled() is the correct function for this but somehow not included in Promise
-  const rates = await Promise.all(
-    exchangeRatePromises.map(promise =>
-      // $FlowExpectedError - Object.values() always produce mixed type so .catch will produce error
-      promise.catch(e => {
-        console.log(e)
-        return 0
-      })
-    )
-  )
-
-  for (let i = 0; i < exchangeRateKeys.length; i++) {
-    const key = exchangeRateKeys[i]
-    const codes = key.split('_')
-    const reverseExchangeRateKey = `${codes[1]}_${codes[0]}${codes[2] ? '_' + codes[2] : ''}`
-    const rate = rates[i]
-    finalExchangeRates[key] = '0'
-    finalExchangeRates[reverseExchangeRateKey] = '0'
 
-    if (rate != null && !isNaN(rate) && rate !== 0) {
-      const rateStr = rate.toFixed(DECIMAL_PRECISION)
-      finalExchangeRates[key] = rateStr
-      finalExchangeRates[reverseExchangeRateKey] = div('1', rateStr, DECIMAL_PRECISION)
+  const serverRates: GuiExchangeRates = {}
+  for (let startIndex = 0; startIndex < exchangeRates.length; startIndex += RATES_SERVER_MAX_QUERY_SIZE) {
+    let endIndex = startIndex + RATES_SERVER_MAX_QUERY_SIZE - 1
+    if (endIndex > exchangeRates.length) {
+      endIndex = exchangeRates.length
     }
+    const query = exchangeRates.slice(startIndex, endIndex)
+    let tries = 5
+    do {
+      const url = pickRandom(RATES_SERVERS) ?? ''
+      const options = {
+        method: 'POST',
+        headers: { 'Content-Type': 'application/json' },
+        body: JSON.stringify({ data: query })
+      }
+      try {
+        const response = await fetch(url + '/v2/exchangeRates', options)
+        if (response.ok) {
+          const json = await response.json()
+          const cleanedRates = asRatesResponse(json)
+          for (const rate of cleanedRates.data) {
+            const { currency_pair: currencyPair, exchangeRate, date } = rate
+            const newDate = new Date(date).valueOf()
+            const now = Date.now()
+
+            if (now - newDate > HOUR_MS) {
+              serverRates[`${currencyPair}_${date}`] = exchangeRate ?? '0'
+            } else {
+              serverRates[currencyPair] = exchangeRate ?? '0'
+            }
+          }
+          break
+        }
+      } catch (e) {
+        console.log(`buildExchangeRates error querying ${url} ${e.message}`)
+      }
+    } while (--tries > 0)
   }
 
-  return finalExchangeRates
-}
+  const exchangeRateKeys = Object.keys(serverRates)
 
-async function fetchExchangeRateHistory(currency: string, date: string): Promise<number> {
-  try {
-    const currencyHistory = await fetch(`https://rates1.edge.app/v1/exchangeRate?currency_pair=${currency}_USD&date=${date}`).catch(e => {
-      console.log('Error fetching fetchExchangeRateHistory', e)
-    })
-    if (currencyHistory != null) {
-      const result = await currencyHistory.json()
-      return parseFloat(result.exchangeRate)
+  for (const key of exchangeRateKeys) {
```

Comment:
Why are you looping over the 'exchangeRateKeys' again?
Can't you set the 'reverseExchangeRate' in the loop on line 85 when you loop over the "cleanedRates.data"?

Reply:
Mostly just copy paste from old implementation